### PR TITLE
Fix 9 year old typo in code that may never have been triggered

### PR DESCRIPTION
**Pull request by josiahcarlson**
Seems like this was never actually hit in reality (`socket.recv()` would have had to return a unicode `str` instead of the typical `bytes`).

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by Mariatta** [comment]
Thanks for the PR. 
`asynchat` has been deprecated since Python 3.6, I'm inclined to not fix this.

**Comment by josiahcarlson** [comment]
Commit log says otherwise: https://github.com/python/cpython/commit/66078ac5d022cf68f5fcd8875649e4e6527648c8

**Comment by josiahcarlson** [comment]
And I just updated my bugs.python.org account information; I signed the CLA a decade ago for the original patch that produced this bad line :P

File: Lib/asynchat.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def handle_read(self):
             return
 
         if isinstance(data, str) and self.use_encoding:
-            data = bytes(str, self.encoding)
+            data = bytes(data, self.encoding)
```
Comment (serhiy-storchaka) [inline Lib/asynchat.py:125]: Would not be better to just remove this code?

Comment (josiahcarlson) [reply Lib/asynchat.py:125]: Arguments for "yes, just remove the code":
1. Code doesn't seem to have ever been triggered in reality
2. If a socket (or socket-like thing) returns a unicode string, that smells like a bug with the socket
3. Implied zens: Errors should never pass silently. Unless explicitly silenced.

Arguments for "no, keep the code and fix it":
1. This was an explicit fix/silencing of an error
2. The intent was to "do the right thing", the existing code can only raise an exception if the clause hits (existing code doesn't do the right thing)
3. Deleting the code will only induce an exception if this case comes up in the future (deleting the code may produce a questionable "right thing")
4. Trivial fix

The real question is whether we want to raise an exception in the case where a socket returns a unicode string, or whether we want to transparently transform the data to bytes with the connection's encoding.

Short version: I think it should be fixed, because that was my intent when I merged the patch 9 years ago.

**Comment by Mariatta** [comment]
Hmm `asynchat` has been documented as deprecated in 3.6: https://docs.python.org/3/library/asynchat.html 
I thought this means that we don't support it anymore.
Maybe I'm wrong about this.

We have an issue tracking replacing `asynchat` with `asyncio` http://bugs.python.org/issue28534
The reason the code is still around is because we haven't got around doing it.

File: Lib/asynchat.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/asynchat.py:125]: I think we need an issue on the bug tracker for discussing. GitHub is used only for discussing technical details of a patch.

**Comment by josiahcarlson** [comment]
Looking at other issues relating to asyncore/asynchat in the bug tracker, many/most have been hanging for years with no resolution. I'll go ahead and close this issue because I suspect all of those will be closed at some point in the future when asyncore/asynchat go away.
